### PR TITLE
Add exception to catch redis connection failure to retry after wait time

### LINKED ISSUE #1153: Add exception catching for when Redis server is down
After my Redis server went down briefly, it seems all of my workers were killed as a result of this error:
```
10:33:55 Worker rq:worker:7870f7200ab64bdfae843ce5f4dc5ae0: found an unhandled exception, quitting...
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/rq/worker.py", line 470, in work
    result = self.dequeue_job_and_maintain_ttl(timeout)
  File "/usr/local/lib/python3.6/site-packages/rq/worker.py", line 521, in dequeue_job_and_maintain_ttl
    job_class=self.job_class)
  File "/usr/local/lib/python3.6/site-packages/rq/queue.py", line 469, in dequeue_any
    result = cls.lpop(queue_keys, timeout, connection=connection)
  File "/usr/local/lib/python3.6/site-packages/rq/queue.py", line 441, in lpop
    result = connection.blpop(queue_keys, timeout)
  File "/usr/local/lib/python3.6/site-packages/redis/client.py", line 1618, in blpop
    return self.execute_command('BLPOP', *keys)
  File "/usr/local/lib/python3.6/site-packages/redis/client.py", line 839, in execute_command
    return self.parse_response(conn, command_name, **options)
  File "/usr/local/lib/python3.6/site-packages/redis/client.py", line 853, in parse_response
    response = connection.read_response()
  File "/usr/local/lib/python3.6/site-packages/redis/sentinel.py", line 55, in read_response
    return super(SentinelManagedConnection, self).read_response()
  File "/usr/local/lib/python3.6/site-packages/redis/connection.py", line 699, in read_response
    response = self._parser.read_response()
  File "/usr/local/lib/python3.6/site-packages/redis/connection.py", line 309, in read_response
    response = self._buffer.readline()
  File "/usr/local/lib/python3.6/site-packages/redis/connection.py", line 241, in readline
    self._read_from_socket()
  File "/usr/local/lib/python3.6/site-packages/redis/connection.py", line 186, in _read_from_socket
    raise ConnectionError(SERVER_CLOSED_CONNECTION_ERROR)
```

I would like them to stay alive, even if they are not working.

Could we add a try/except block around lines like: https://github.com/rq/rq/blob/75644ba9486ef3d8b17e8b9cf4dac91fab552375/rq/queue.py#L441

### PR DESCRIPTION
**Pull request by Asrst**
Resolves #1153
Could be also potential fix for #998

formatted the code following the comments made by @selwin  on #1261 . Also, added the `max_connection_wait_time` to limit max retry timeout to 60 seconds.

### CODE REVIEW

File: rq/worker.py [hunk 1]
```diff
@@ -591,6 +594,15 @@ def dequeue_job_and_maintain_ttl(self, timeout):
                 break
             except DequeueTimeout:
                 pass
+            except redis.exceptions.ConnectionError as conn_err:
+                self.log.error('Could not connect to Redis instance: %s Retrying in %d seconds...', 
+                                conn_err, connection_wait_time)
+                time.sleep(connection_wait_time)
+                connection_wait_time *= self.exponential_backoff_factor
+                if connection_wait_time > self.max_connection_wait_time:
+                    connection_wait_time = self.max_connection_wait_time
```
Comment (selwin) [inline rq/worker.py:603]: Nitpick: `connection_wait_time = min(connection_wait_time, self.max_connection_wait_time)`

Comment (Asrst) [reply rq/worker.py]: resolved it. thanks for the suggestion.

**Comment by selwin** [comment]
Thanks!

**Comment by vincentwoo** [comment]
Would really love a release cut with this! One of our last big rq issues is resiliency to redis dying as the worker pulls a job (we lose the task entirely currently).

**Comment by jmgirven** [comment]
Shouldn't the `except DequeueTimeout` also reset the `connection_wait_time`? It seems that except is being hit when successfully connected to redis, but no jobs are there to process. So a case like:

- Successfully connected to redis
- Lost connection to redis for 10 minutes (`connection_wait_time` reaches `max_connection_wait_time`)
- Reconnected to redis & stays connected for hours, but has no jobs to process
- Lost connection to redis (`connection_wait_time` is immediately at `max_connection_wait_time`)

Is possible, but seems wrong. @Asrst

**Pull request merged**